internal/pkg/service/buffer/storage/local/volume/volume.go

```diff
@@ -0,0 +1,210 @@
+// Package volume contains common code for reader.Volumes and writer.Volumes implementations.
+// The OpenVolumes function detects and opens all volumes in the volumesPath.
+//
+// Volume relative path has the following format: "{type}/{label}".
+//
+// The type is later used when assigning volumes.
+// Different use-cases may prefer a different type of volume.
+//
+// The label has no special meaning, volumes are identified by the unique storage.VolumeID,
+// which is read from the IDFile on the volume, if the file does not exist,
+// it is generated and saved by the writer.Volume.
+package volume
+
+import (
+	"io/fs"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/log"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/storage"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
+)
+
+const IDFile = "volume-id"
+
+// Volume instance common interface.
+type Volume interface {
+	Path() string
+	Type() string
+	Label() string
+	ID() storage.VolumeID
+	Close() error
+}
+
+// Volumes manages volumes in the path, each instance has V type.
+type Volumes[V Volume] struct {
+	logger  log.Logger
+	path    string
+	factory Opener[V]
+	byID    map[storage.VolumeID]V
+	byType  map[string][]V
+}
+
+// Opener opens volume instance of the V type.
+type Opener[V Volume] func(info Info) (V, error)
```

Comment:
Puziva sa genericke programovanie, `V` musi implementovat dane `Volume` metody, ... realne je to potom `writer.Volume` alebo `reader.Volume`.